Complete refactoring (#17)
openfun/xblock-configurable-lti-consumer: README.md

# Configurable LTI Consumer XBlock

> This XBlock builds on top of [edX LTI consumer

Reviewer: ```suggestion
This XBlock is built on top of [edX LTI consumer
```

Xblock](https://github.com/edx/xblock-lti-consumer) to ease its configuration by Instructors.

Here are a few examples of what this module brings to Open edX:

> - Pre-configure LTI services that can be added in one-click from the `Advanced` component button,

Reviewer: ```suggestion
- Pre-configured LTI services that can be added in one-click from the `Advanced` component button,
```

> - Enforce configuration for some LTI Services by targetting a launch url pattern,

Reviewer: ```suggestion
- Enforced configuration for some LTI Services by targetting a launch url pattern,
```

> - Enforce a default configuration for all LTI Services for better control, security or ergonomy,

Reviewer: ```suggestion
- Enforced default configuration for all LTI Services for better control, security or ergonomy,
```

- Platform-level LTI credentials configuration available to all courses.

>   courses.

Reviewer 2: `courses.` is written two times

Author: arf

> - Keep LTI credentials at the level of the platform configuration and make them available to all

Reviewer: ```suggestion
- Platform-level LTI credentials configuration available to all courses.
```

            "launch_url": "https://marsha.education/lti-video/",
            "lti_id": "marsha",
        },
    },
    {
        "pattern": ".*ltiapps\\.net.*",
        "hidden_fields": ["launch_target"],
        "defaults": {"launch_target": "modal"},
    },
    {
        "display_name": "LTI consumer",
        "pattern": ".*",
        "hidden_fields": ["ask_to_send_username", "ask_to_send_email"],
        "defaults": {
            "ask_to_send_email": True,
            "launch_target": "new_window",
            "ask_to_send_username": True,
        },
    },
]
```

This configuration does several things:

- it adds a "Marsha Video" link behind the `Advanced` component button in the Studio to add a
  video in one click. The video is automatically added to the
  [Marsha](https://github.com/openfun/marsha) instance pointed by the launch url,
- it forces all LTI consumer XBlocks that are pointing to `ltiapps.net` to use a modal,
- it proposes all other LTI consumer XBlocks to open in a new window as a default and forces
  to ask before sending the user's username and email.

The order of each configuration in this list is important because we will use, for a given

> XBlock, the first configuration pattern that matchs its launch url.

Reviewer: ```suggestion
XBlock, the first configuration pattern that matches its launch url.
```

Note that the workbench included in the present repository is running this configuration

> (see ./config/settings.yml.dist) on the official France Université Numérique

Reviewer: ```suggestion
(see [config/settings.yml.dist](./config/settings.yml.dist)) on the official France Université Numérique
```

[Open edX extended Docker image](https://github.com/openfun/openedx-docker).

## Integration to Open edX Studio

For now, this project requires a small fork of [edx/edx-platform](github.com/edx/edx-platform)
if you want to add, for some of your configurations, preconfigured options in the
`Advanced` component button of the Studio.

In our opinion, this small fork is worth applying because:

> - it will save you from installing many XBlocks and help keep your Open edX installation manageable,

Reviewer: ```suggestion
- it will save you from installing many XBlocks and helps to keep your Open edX installation manageable,
```

WDYT?

Author: will save you ... and will help you. But your proposal also works so let's change.

```

Once the project has been cloned on your machine, you will need to build a
custom [edx-platform docker image](https://github.com/openfun/openedx-docker) that
includes the configurable LTI consumer XBlock and setup a development environment that
includes all required services up and running (more on this later):

```bash
$ cd xblock-configurable-lti-consumer
$ make bootstrap
```

If everything went well, you should now be able to access the following services:

- Open edX LMS: http://localhost:8072
- Open edX CMS: http://localhost:8082

with the following credentials:

```
email: admin@example.com
password: admin
```

## Configuration examples

The following LTI configuration does several things:

- it adds a "Marsha Video" link behind the `Advanced` component button in the Studio to add a
  video in one click. The video is automatically added to
  [Marsha](https://github.com/openfun/marsha),
- it forces all LTI consumer XBlocks that are pointing to `ltiapps.net` to use a modal, together
  with a few other settings,
- it forces all other LTI consumer XBlocks to open in a new window and ask before sending the

>   user's username and email.

Reviewer: IMO this explanation should be placed after the example. WDYT?

```yaml

> LTI_XBLOCK_CONFIGURATIONS:

Reviewer: Add a comment with an example path for this file content.

  # A configuration that adds a Marsha Video XBlock
  - boilerplate: marsha
    display_name: Marsha Video
    oauth_consumer_key: InsecureOauthConsumerKey
    shared_secret: InsecureSharedSecret
    fields:
      lti_id: marsha

>       description: ""

Reviewer: Does it worth it to add a short description here?

Author: yes to hide this useless field!

Author: We have finally decided to split the list of hidden fields from the list of values for a field...

      launch_target: iframe
      launch_url: https://marsha.education/lti-video/
      custom_parameters: []
      button_text: button
      inline_height: 100
      modal_height: 400
      modal_width: 80
      has_score: False
      weight: 0
      hide_launch: False
      accept_grades_past_due: False
      ask_to_send_username: True
      ask_to_send_email: True
  # A configuration that applies to all LTI consumer XBlocks pointing to a launch url on the
  # ltiapps.net domain. It does not add a link to the `Advanced` Component button in the Studio
  - pattern: .*ltiapps\.net.*
    fields:
      launch_target: modal
      custom_parameters: []
      button_text: LTI demo
      inline_height: 100
      has_score: False
      weight: 0
      hide_launch: False
      accept_grades_past_due: False
      ask_to_send_username: True
      ask_to_send_email: True
  # A configuration that enforces a default behavior for all LTI consumer XBlocks
  - pattern: .*
    fields:
      launch_target: in_window
      ask_to_send_username: True
      ask_to_send_email: True

> 

Reviewer: Extra empty row.

Author: I replaced the whole thing by a Python version as suggested by @rmoch

> ```

Reviewer: Can you mention in which file we expect this configuration to be?

Reviewer 2: also, why yaml ? most edx developer would create such settings in Python

Author: done

The order of each configuration in this list is important because we will use, for a given

> XBlock, the first configuration that matchs its launch url.

Reviewer: ```suggestion
XBlock, the first configuration pattern that matches its launch url.
```

@@ -27,6 +27,15 @@ from xblock_django.models import XBlockStudioConfigurationFlag
 from xmodule.modulestore.django import modulestore
 from xmodule.modulestore.exceptions import ItemNotFoundError
 
+# Try to import the configurable LTI consumer xblock
+try:
+    from lti_consumer import add_dynamic_components
+except ImportError:
+    add_dynamic_components = None
+
+
 __all__ = [
     'container_handler',
     'component_handler'
@@ -359,6 +368,17 @@ def get_component_templates(courselike, library=False):
         "support_legend": create_support_legend_dict()
     }
     advanced_component_types = _advanced_component_types(allow_unsupported)
+
+    # Read configuration setting to create specific buttons
+    if add_dynamic_components:
+        add_dynamic_components(
+            getattr(settings, "LTI_XBLOCK_CONFIGURATIONS", []),
+            advanced_component_templates,
+            categories,
+            create_template_dict,
+            course_advanced_keys
+        )
+
     # Set component types according to course policy file
     if isinstance(course_advanced_keys, list):
         for category in course_advanced_keys:

> ```

Reviewer: We can maybe store this patch as a gist an point to it in the documentation?

Author: so remove it from the documentation?

Reviewer: Yes.

Author: Done
